core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,466 @@
+                if (addedIndex < currentTombstones.size()) {
+                    added = currentTombstones.subList(addedIndex + 1, currentTombstones.size());
+                } else {
+                    added = Collections.emptyList();
+                }
+                // look from the front for the removed tombstones
+                final Tombstone firstTombstone = currentTombstones.get(0);
+                int idx = previousTombstones.indexOf(firstTombstone);
+                if (idx < 0) {
+                    // the first tombstone in the current list wasn't found in the previous list,
+                    // which means all tombstones from the previous list have been deleted.
+                    assert added.equals(currentTombstones); // all previous are removed, so the current list must be the same as the added
+                    idx = previousTombstones.size();
+                }
+                removed = idx;
+            }
+            this.added = Collections.unmodifiableList(added);
+            this.removedCount = removed;
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            out.writeList(added);
+            out.writeVInt(removedCount);
+        }
+
+        @Override
+        public IndexGraveyard apply(final MetaData.Custom previous) {
+            @SuppressWarnings("unchecked")
+            final IndexGraveyard old = (IndexGraveyard) previous;
+            if (removedCount > old.tombstones.size()) {
+                throw new IllegalStateException("IndexGraveyardDiff cannot remove " + removedCount + " entries from " +
+                                                old.tombstones.size() + " tombstones.");
+            }
+            final List<Tombstone> newTombstones = new ArrayList<>(old.tombstones.subList(removedCount, old.tombstones.size()));
+            for (Tombstone tombstone : added) {
+                newTombstones.add(tombstone);
+            }
+            return new IndexGraveyard.Builder().addBuiltTombstones(newTombstones).build();
+        }
+
+        /** The index tombstones that were added between two states */
+        public List<Tombstone> getAdded() {
+            return added;
+        }
+
+        /** The number of index tombstones that were removed between two states */
+        public int getRemovedCount() {
+            return removedCount;
+        }
+    }
+
+    /**
+     * An individual tombstone entry for representing a deleted index.
+     */
+    final public static class Tombstone implements ToXContent, Writeable<Tombstone> {
+
+        private static final String INDEX_KEY = "index";
+        private static final String DELETE_DATE_IN_MILLIS_KEY = "delete_date_in_millis";
+        private static final String DELETE_DATE_KEY = "delete_date";
+        private static final ObjectParser<Tombstone.Builder, Void> TOMBSTONE_PARSER =
+            new ObjectParser<>("tombstoneEntry", Tombstone.Builder::new);
```

Comment:
I think I'd prefer it if the construction was moved to be alongside the initialization in the class initializer.